node/src/reactor/joiner.rs

```diff
@@ -870,15 +882,23 @@ impl reactor::Reactor for Reactor {
             }
 
             Event::LinearChainAnnouncement(LinearChainAnnouncement::BlockAdded(block)) => {
-                let block_hash = *block.hash();
-                reactor::wrap_effects(
+                let mut effects = reactor::wrap_effects(
                     Event::EventStreamServer,
                     self.event_stream_server.handle_event(
                         effect_builder,
                         rng,
-                        event_stream_server::Event::BlockAdded { block_hash, block },
+                        event_stream_server::Event::BlockAdded(block.clone()),
                     ),
-                )
+                );
+                let reactor_event =
+                    Event::LinearChainSync(linear_chain_sync::Event::BlockHandled(block.clone()));
+                effects.extend(self.dispatch_event(effect_builder, rng, reactor_event));
+                let reactor_event =
+                    Event::ChainspecLoader(chainspec_loader::Event::CheckForNextUpgrade);
+                effects.extend(self.dispatch_event(effect_builder, rng, reactor_event));
+                let reactor_event = Event::Consensus(consensus::Event::BlockAdded(block));
+                effects.extend(self.dispatch_event(effect_builder, rng, reactor_event));
+                effects
```

Comment:
I really dislike this – I think it's wrong to put the domain logic here and it's error-prone.